instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java

```diff
@@ -0,0 +1,154 @@
+  @CsvSource({"by instance", "by class", "by name"})
+  void shouldReplaceHandler(String desc) {
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = channel.pipeline();
+    HttpClientCodec httpHandler = new HttpClientCodec();
+
+    // remove the default head and tail handlers
+    if (channelPipeline.first() != null) {
+      channelPipeline.removeFirst();
+    }
+    if (channelPipeline.last() != null) {
+      channelPipeline.removeLast();
+    }
+
+    // no handlers initially
+    assertEquals(0, channelPipeline.toMap().size());
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addFirst("test", noopHandler);
+
+    // only the noop handler
+    assertEquals(1, channelPipeline.toMap().size());
+    assertEquals(noopHandler, channelPipeline.first());
+
+    if ("by instance".equals(desc)) {
+      channelPipeline.replace(noopHandler, "http", httpHandler);
+    } else if ("by class".equals(desc)) {
+      channelPipeline.replace(noopHandler.getClass(), "http", httpHandler);
+    } else if ("by name".equals(desc)) {
+      channelPipeline.replace("test", "http", httpHandler);
+    }
+
+    // noop handler was removed; http and instrumentation handlers were added
+    assertEquals(2, channelPipeline.toMap().size());
+    assertEquals(httpHandler, channelPipeline.first());
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+
+    NoopChannelHandler anotherNoopHandler = new NoopChannelHandler();
+    channelPipeline.replace("http", "test", anotherNoopHandler);
+
+    // http and instrumentation handlers were removed; noop handler was added
+    assertEquals(1, channelPipeline.toMap().size());
+    assertEquals(anotherNoopHandler, channelPipeline.first());
+  }
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4056
+  @Test
+  void shouldAddAfterAndRemoveLastHandler() {
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = channel.pipeline();
+    HttpClientCodec httpHandler = new HttpClientCodec();
+
+    if (channelPipeline.first() != null) {
+      channelPipeline.removeFirst();
+    }
+    if (channelPipeline.last() != null) {
+      channelPipeline.removeLast();
+    }
+
+    assertEquals(0, channelPipeline.toMap().size());
+
+    channelPipeline.addLast("http", httpHandler);
+    assertEquals(2, channelPipeline.toMap().size());
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last().getClass().getSimpleName(), "HttpClientTracingHandler");
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addAfter("http", "noop", noopHandler);
+
+    assertEquals(3, channelPipeline.toMap().size());
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last(), noopHandler);
+
```

Comment:
as far as I can tell there was a bit more code here in the groovy test

Reply:
updated. out of sync from main branch.